### PR TITLE
Add blogpost on dereference pushdown

### PR DESCRIPTION
**Pull request by phd3**

### CODE REVIEW

**Comment by martint** [comment]
cc @mosabua

**Comment by mosabua** [comment]
Can you reformat the source to 80 char width please. This will make the review much easier.

**Comment by mosabua** [comment]
Whatever formatting tool you used @phd3 .. it broke the words apart. Please fix

**Comment by phd3** [comment]
@mosabua yeah :/ do you know of a good tool to do this automatically?

**Comment by phd3** [comment]
@mosabua addressed your comment.

File: _posts/2020-08-11-dereference-pushdown.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
+---
+layout: post
+title:  "Faster Queries on Nested Data"
+author: Pratham Desai
+---
+
+Significant performance improvements have been added in PrestoSQL version 334
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:7]: Make it active voice and present tense.

Presto version 334 adds significant performance improvements for queries...

File: _posts/2020-08-11-dereference-pushdown.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
+---
+layout: post
+title:  "Faster Queries on Nested Data"
+author: Pratham Desai
```
Comment (martint) [inline _posts/2020-08-11-dereference-pushdown.md:4]:
```suggestion
author: Pratham Desai, LinkedIn
```

File: _posts/2020-08-11-dereference-pushdown.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+from existing flows on row-oriented formats. 
+
+The dereference pushdown optimization in Presto is having a massive impact on 
+our ingestion story. Nested data is now being made available faster for 
+consumption, while maintaining performance parity for analytical queries.
+
+
+# Example
+
+Say we have a hive table `jobs`, with a struct-typed column `job_info` in the
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:44]: Hive table

File: _posts/2020-08-11-dereference-pushdown.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+The dereference pushdown optimization in Presto is having a massive impact on 
+our ingestion story. Nested data is now being made available faster for 
+consumption, while maintaining performance parity for analytical queries.
+
+
+# Example
+
+Say we have a hive table `jobs`, with a struct-typed column `job_info` in the 
+schema. Say the column `jobs` is wide and deeply nested, i.e. `ROW(company
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:45]: Drop "Say"

File: _posts/2020-08-11-dereference-pushdown.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+With dereference pushdown, Presto optimizes queries by extracting the sufficient
+ fields from a `ROW` as early as possible. This is enforced by modifying the 
+query plan through a set of optimizers, and can be broadly divided into two 
+parts.
+
+First, dereference projections are extracted in the query plan and pushed as 
+close to the table scan as possible. This happens independent of what the 
+connector is. Secondly, there is a further improvement for Hive tables. The 
+Hive Connector and ORC/Parquet readers have been optimized to scan only the
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:74]: Hive connector

File: _posts/2020-08-11-dereference-pushdown.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+query plan through a set of optimizers, and can be broadly divided into two 
+parts.
+
+First, dereference projections are extracted in the query plan and pushed as 
+close to the table scan as possible. This happens independent of what the 
+connector is. Secondly, there is a further improvement for Hive tables. The 
+Hive Connector and ORC/Parquet readers have been optimized to scan only the 
+sufficient subfield columns.
+
+Pushdown of predicates on the subfields is also a crucial optimization. i.e. if
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:77]: Replace i.e. with "For example, "

File: _posts/2020-08-11-dereference-pushdown.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+sufficient subfield columns.
+
+Pushdown of predicates on the subfields is also a crucial optimization. i.e. if 
+a query has filters on subfields (i.e. `a.b > 5`), they should be utilized by 
+ORC/Parquet readers while scanning files. The pushdown helps with the pruning 
+of files, stripes and row-groups based on column-level statistics. This 
+optimization is achieved as a byproduct of the above two optimizations.
+
+With the dereference pushdown, queries observe significant performance gains in 
+terms of cpu/memory usage and query runtime, roughly proportional to the
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:84]: CPU

File: _posts/2020-08-11-dereference-pushdown.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+optimization is achieved as a byproduct of the above two optimizations.
+
+With the dereference pushdown, queries observe significant performance gains in 
+terms of cpu/memory usage and query runtime, roughly proportional to the 
+relative size of nested columns compared to the accessed fields.
+
+## Pushdown in Query Plan
+
+The goal here is to execute dereference projections as early as possible. This 
+usually means performing them right after the table scans in most cases.
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:90]: drop either "usually" or "in most cases"

File: _posts/2020-08-11-dereference-pushdown.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+pushdown is enabled.
+
+![](/assets/blog/dereference-pushdown/transformed_plan.png)
+
+The projection `job_info.company` now directly follows the scan of `jobs` table,
+ avoiding the propagation the `job_info` through `Limit` and `Join` nodes. Note 
+that all of `job_info` is still being scanned, and pruning it in the reader 
+requires connector-dependent optimizations.
+
+## Pushdown in Hive Connector
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:106]: Pushdown in the Hive Connector

File: _posts/2020-08-11-dereference-pushdown.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+fashion even for subfields. If we have a column `STRUCT(f1, f2, f3)`, the 
+subfields `f1`, `f2` and `f3` are stored as independent columns. An optimized 
+query engine should only scan the required fields through its ORC reader, 
+skipping the rest. This optimization has been added for Hive connector. 
+
+Dereference projections above a `TableScanNode` are pushed down in the Hive 
+connector as "virtual" (or “projected”) columns. The query plan is modified to 
+refer to these new columns. For the query `Q`, `jobs` table would be scanned 
+differently with this optimization, as shown below. The projection is now 
+embedded in the Hive Connector. Here, `job_info#company` can be thought of as
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:118]: Hive connector

File: _posts/2020-08-11-dereference-pushdown.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+Dereference projections above a `TableScanNode` are pushed down in the Hive 
+connector as "virtual" (or “projected”) columns. The query plan is modified to 
+refer to these new columns. For the query `Q`, `jobs` table would be scanned 
+differently with this optimization, as shown below. The projection is now 
+embedded in the Hive Connector. Here, `job_info#company` can be thought of as 
+a virtual column representing the subfield `job_info.company`. 
+
+![](/assets/blog/dereference-pushdown/connector_pushdown.png)
+
+Hive connector handles the projections before returning columns to Presto's
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:123]: The Hive connector

File: _posts/2020-08-11-dereference-pushdown.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+to other connectors supporting nested types.
+
+## Conclusion
+
+Pushing down dereference operations in the query provides massive performance 
+gains, especially while operating on large structs. At LinkedIn and Lyft, this 
+feature has shown great impact for analytical queries on nested datasets. 
+
+We're excited for the community to try it out. Please feel free to reach out on 
+[slack](https://join.slack.com/t/prestosql/shared_invite/enQtNTMyNzU2NzQ1NzQ4LWU
```
Comment (mosabua) [inline _posts/2020-08-11-dereference-pushdown.md:201]: change link to [Slack](/slack.html)

**Comment by phd3** [comment]
@martint @mosabua resolved your comments. If it looks good, I can squash the fixup.

**Comment by martint** [comment]
Good stuff! I squashed the commits and merged it. Thanks!

**Pull request merged**